0000-ilp-over-http.md

```diff
@@ -0,0 +1,59 @@
+---
+title: ILP Over HTTP
+draft: 1
+---
+
+# ILP Over HTTP
+> A bilateral communication protocol for server-to-server connections
+
+## Motivation
+
+Scaling Interledger infrastructure to handle large volumes of ILP packets requires horizontally scaling connectors. Using HTTP for bilateral communication enables service providers to leverage standard tools and services for hosting, load balancing, Distributed Denial of Service (DDoS) protection, and monitoring.
+
+## Overview
+
+In an ILP Over HTTP connection, both peers run HTTP servers with accessible HTTPS endpoints. When peering, the peers exchange their respective URLs, authentication tokens or TLS certificates, ILP addresses, and settlement-related details.
+
+Each ILP Prepare packet is sent as the body of an HTTP request to the peer's server endpoint. ILP Fulfill or Reject packets are returned as the body of the HTTP response.
+
+## Specification
+
+This is a minimal protocol built on HTTP. HTTP/2 is RECOMMENDED for performance reasons, although HTTP/1.1 MAY also be used. Implementations SHOULD support HTTP version negotiation via Application Protocol Negotiation (ALPN).
+
+### Authentication
+
+Peers MAY use any standard HTTP authentication mechanism to authenticate incoming requests. Bearer Tokens or TLS Client Certificates are RECOMMENDED.
```

Comment:
With HTTP2 header compression, it wouldn't actually send the full string over the wire each time. 

Would using TLS client certs do what you're suggesting? If so, that sounds good to me. However, as far as I understand, client certs can be somewhat inconvenient to setup so I'd prefer to give options and let people determine what makes sense for their setup.

Reply:
Client certs are definitely harder to setup but they'd be more efficient.

My point here is to avoid suggesting bearer tokens or using them in the example because the result will be that it becomes the de-facto standard for this protocol.

Reply:
> the result will be that it becomes the de-facto standard for this protocol

I don't think that's true. As far as I know, most HTTP libraries support both methods and peers will use whatever setup makes sense for them.

Reply:
Peers could have agreed to auth BTP using proper WebSocket handshakes but they all settled on credentials in the URL translated into an `auth` subprotocol because that's what was in the examples...

Reply:
Actually the BTP spec is pretty specific about how it expects peers to authenticate one another, and it doesn't make it sound like there are options:

> Before anything else, when a client connects to a server, it sends a special Message request. Its primary protocolData entry MUST have name 'auth', content type MIME_APPLICATION_OCTET_STREAM, and empty data, and among the secondary entries, there MUST be a UTF-8 'auth_token' entry. The further secondary protocol data entries of this Message request MAY also be used to send additional information to the server. In situations where no authentication is needed, the 'auth_token' data can be set to the empty string, but it cannot be omitted.

> **If the client sends any BTP call that is not a Message, or sends a Message call whose primary sub-protocol is not auth, the server should respond with an Error and close the connection.** [emphasis added]